### ISSUE TITLE
Upcoming CSS changes

### ISSUE THREAD

**Issue by SkyCrawl** [gollum/gollum#917]
Issue in accordance with:  #906 , #908, #909, #910 

@sunny @dometto @bartkamphorst 

I've had a look at Gollum's styles and it's a **much** larger and more complex system than I thought it would be. From my humble point of view, it's way too large and complex (for the more or less simple infrastructure and looks of Gollum's pages) and stands on the edge of being impossible to maintain. Newcomers like me can only stare from afar :p.
This could **definitely** use a **lot of** help from SASS :). It would probably take a major amount of work to rewrite all the styles into SASS and get it all working again but every second would count greatly towards dev-experience and future maintenance. On top of that, any potential templating would become so much easier. Inspiration may be taken from:
- [Vaadin's article on SASS](https://vaadin.com/book/vaadin7/-/page/themes.sass.html)
- [Theming with SASS](http://morevaadin.com/content/easy-theming-with-valo/)
- [Example of two 10-minute themes using a nice SASS theming engine](http://jaxenter.com/why-youll-want-to-check-out-vaadin-7-3s-new-valo-theme-engine-108048.html)
- [Ruby SASS - compile all SAS files into a single CSS to be included in pages](https://www.npmjs.org/package/gulp-ruby-sass)

While I can incorporate the changes in `template.css` and `print.css` in this particular case, I have no idea whether that's actually OK with the rest of the styles and honestly, don't really want to be finding out... That's as far as my contributions in this regard will go, unless we (or me, possibly) "sit down" to rewrite all this first :).
Some changes had, in my opinion, best be accompanied by mustache template changes:
- Page title should only duplicate `<h1>`'s styles and another tag be used for it. This is so that `<h1>` may be used in pages themselves and header numbering will not take any interference from the page title. Unless, of course, the associated "trigger" style name would be pushed further in the DOM (not `<body>`) so as not to encompass the page title.
- While applying `position: absolute;` to menu actions should not be problematic, I think that with regard to the look we want to achieve, page title tag should follow (not precede) the action menu.
- If header numbering is used and table of contents displayed on a page, they should be in sync (currently, table of contents is not being numbered). CSS-wise, there shouldn't be much of a problem but can it be easily handled from your end?

**Comment by dometto** [comment]
@SkyCrawl, does any more fundamental stuff need to happen before you can implement the changes mentioned in the top post (when you have the time, of course)?

**Comment by SkyCrawl** [comment]
Once all the original styles are integrated and used, and when I'm given instructions about how to develop in this new system, we're all set I suppose. After that, I just need to get familiar with the new design and SASS as a whole...
Note, however, that I won't be able to start with this for at least a month yet and when/if I do, I intend to take it slow :).

**Comment by techwiz24** [comment]
Can we consider changing the maximum width of pages so it's not hard-coded?

I apply a patch with this in it after every gollum update to do this for my wiki's, @dometto requested I post it here also (coming from #945)

```diff
   #wiki-wrapper {
-    max-width: 920px;
+    max-width: 95%;
     padding-left:20px;
     padding-right:20px;
   }
```

I'm no designer, but 95% seems to work pretty well for me :)

**Comment by SkyCrawl** [comment]
@techwiz24 
Yes, percentage values work well for a single computer/monitor but are most likely not a good choice for massive use. For example, 95% would make the wiki's text so wide on my notebook and external monitor that it would become really tiresome to read. And all the more irritating if you're looking for a specific information somewhere in the text or when you're trying to "skip-read"... Should Gollum act that way, I would personally override it in custom CSS to something like `max-width: 920px;`. It's alright to adjust the page width for yourself but you probably shouldn't push it to (potentially) everyone else who might not even know how to shrink it again :).

Width is, in my opinion, usually hardcoded because one can not predict what the value will actually translate into otherwise. If your page's design will not be broken in some corner cases, it is most likely interfering with the goals of webdesign. Human vision has a certain angle (some kind of mean value) in which it can process information well (e.g. with low effort) after all... Especially when concentrating and struggling to understand the text. While thinner pages tend to be cluttered quite a bit, they also allow readers to process/use them better as a whole :).
**Long story short**: "don't use large widths for pages designated to be read".

If you really wish for Gollum's width to be dynamic (and if you're lazy to manually override it after each page load):
* Override Gollum's CSS, just as you have shown in your example, but only as long as you're using Gollum by yourself or you know that other readers won't mind.
* If it's for your personal use, you might try to press for a special URL parameter, cookie or HTML5 store value.

**Comment by techwiz24** [comment]
@SkyCrawl Thanks for the feedback! That's why I'm not a web developer :)

I was just adding my 2cents because I thought it felt weird having tons of whitespace on the sides of the content.

**Comment by xu4wang** [comment]
@techwiz24 thanks for sharing the tips. 95% works for me too.  Also I figure it out that I need to use --css to apply the change

**Comment by SkyCrawl** [comment]
@rtrvrtg I can finally join you in the efforts, once I setup a dev environment this evening :). How is the conversion going? Do you have plans to continue? And what is the current status on SASS compilation? Is it real-time (when Sinatra starts, I suspect) or do I need to compile it manually?

**Comment by bartkamphorst** [comment]
Good news, @SkyCrawl!

**Comment by aldanor** [comment]
Looking forward to this :+1: Will be glad to help once there's some working version.

**Comment by dometto** [comment]
@aldanor, any help is much appreciated! There is already a sass to css branch, but maybe there's more goodies from @rtvg coming up. :)

**Comment by SkyCrawl** [comment]
> @aldanor, any help is much appreciated! There is already a sass to css branch, but maybe there's more goodies from @rtvg coming up. :)

I doubt it, @dometto . At least for the time being.

> How is the conversion going? Do you have plans to continue? And what is the current status on SASS compilation? Is it real-time (when Sinatra starts, I suspect) or do I need to compile it manually?

Let's answer this, too:
* SASS has been successfully integrated and working structural base has been created.
* Changes are hot-deployed on browser refresh. No manual procedures whatsoever.

> Looking forward to this :+1: Will be glad to help once there's some working version.

At this moment, there is one [here](https://github.com/SkyCrawl/gollum/tree/css-to-sass), @aldanor . You're welcome to join. Before you do, however, please view [this](https://github.com/SkyCrawl/gollum/blob/css-to-sass/TODO.md) file. There's still a lot to do, so let's agree on some basic rules and the level we want to achieve, and divide the work. I'll make you a contributor whenever you wish :).

----

@dometto @bartkamphorst @sunny 
Could I ask somebody to:
* Compile a list of issues related to CSS/SASS and post them in [this](https://github.com/SkyCrawl/gollum/blob/css-to-sass/TODO.md) file (once I make them a contributor of course)? If so, would you be willing to add new issues as they appear?
* Alternatively, could you tag the issues (in here) with something like `style` or `css`?

**Comment by dometto** [comment]
@SkyCrawl, thanks for taking the initiative here. As I mentioned I'll be a bit occupied this month, but I can definitely help compile a list of style related issues. I'll get on it this weekend.

**Comment by SkyCrawl** [comment]
Okay, thanks @dometto and @bartkamphorst :)

**Comment by dometto** [comment]
@bartkamphorst beat me to it. ty!

**Comment by SkyCrawl** [comment]
@dometto @bartkamphorst @sunny 
Further consolidation of CSS into SASS is going well. Although I've been busy getting familiar with Gollum's internals more than I had hoped for, I managed to update and simplify quite a bit so far and things should only get faster if I maintain the current pace. Details can be viewed at the bottom of my [state-tracking file](https://github.com/SkyCrawl/gollum/blob/css-to-sass/TODO.md).

In the course of my adventure, I discovered many messy and incomplete parts of Gollum, most of which should probably be made into issues. Details can again be viewed in the above file. What I would like to discuss in regard to my work at this moment, however, is the following exerpt:
```
1. Test Gollum repositories should (IMHO) be published as branches.
2. Introduce AJAX - useful for (AT LEAST) submitting actions (without forms) and error handling.
3. Merge `Pages` and `File view` pages.
4. Merge `History` and `Latest changes` pages?
5. Code editor & livepreview - upgrade, consolidate and simplify.
```
1. Test repositories
     * The fact that test repositories are included in the "server-form" kind of made me wonder why. I suspect it had to do with the simplicity and performance of tests but on the other hand, it adds complexity to editing, updating them and displaying any kind of information about them. Imho, uploading them as branches to Gollum while (possibly) including them as submodules would be a much better solution - nice & easy updating via GitHub's interface and always up-to-date clones in dev environment if we force pull before tests are executed :).
    * Why are there so many? Wouldn't it be nicer if we had a single test repository that (say) defines all test cases in separate folders?
2. I think I saw a hint of a future "to-AJAX-shift" in #826 when skip-reading it. The above points may also be good reasons to do it.
3. This would probably be quite a bit simpler with AJAX enabled but what I'd like to ask is why wasn't a unified page/file view an option (e.g. a checkbox that enables/disables page-only view)? Speaking from the perspective of a "CSS guy", they're the same and I'd like to unify them as soon as possible :).
4. Pretty much the same question as in point 3 - they seem the same to me and should be unified.
5. I'll definitely read #826 some time in the future but I'd still like to ask somebody to create and maintain a detailed roadmap for Gollum, if only to stop people from asking "silly" questions: is there a plan to revisit code and live preview editors soon? :laughing:

**Comment by bartkamphorst** [comment]
Great work, @SkyCrawl ! I'm very excited about all the improvements you are making. Quick response RE 4: as I recall `latest_changes.mustache` is a simplified version of `history.mustache`. If you think they're better unified, I'm all for it!

**Comment by SkyCrawl** [comment]
@dometto @bartkamphorst @sunny 
How is it with thread-safety in Gollum? Does the current chain of `WEBRick<->Rack<->Sinatra` use threads to serve incoming requests? Is the `Gollum::Wiki` object thread-safe?

**Comment by dometto** [comment]
@SkyCrawl: sinatra itself is thread-safe, but doesn't impose its own concurrency--that depends on the webserver you use it with (see [here](http://stackoverflow.com/questions/6278817/is-sinatra-multi-threaded)). Concerning the `Gollum::Wiki` object I have to say I just don't know. It probably depends on the underlying git layer. Rugged is [supposedly thread-safe](http://stackoverflow.com/questions/29250745/are-rugged-repository-instances-threadsafe).

**Comment by SkyCrawl** [comment]
Thank you, @dometto . I think that when I'm done with all of my changes, Gollum will be thread safe and available for public use if:
* the underlying git layer is thread-safe,
* sinatra routes lock resources when updating them.
But Gollum is obviously designed to be a personal Wiki anyway so this doesn't have much priority, at least for now :). I was asking mainly because I wanted to get rid of creating new WIki instances all the time, in which I think I already succeeded.

**Comment by dometto** [comment]
Ah, yes -- notice that Rugged is only thread-safe when not re-using the same `Repository` instance, which will probably happen when you don't re-create the `Wiki` object for each request.

**Comment by dometto** [comment]
Closing in favor of #1243.
